src/connectors/impls/clickhouse/conversion.rs

```diff
@@ -0,0 +1,457 @@
+            // provides is incorrect.
+            let inner_value = convert_value(column_name, value, expected_inner_type)?;
+            Ok(CValue::Nullable(Either::Right(Box::new(inner_value))))
+        }
+
+        DummySqlType::UInt8 => get_and_wrap(context, ValueAccess::as_u8, CValue::UInt8),
+
+        DummySqlType::UInt16 => get_and_wrap(context, ValueAccess::as_u16, CValue::UInt16),
+
+        DummySqlType::UInt32 => get_and_wrap(context, ValueAccess::as_u32, CValue::UInt32),
+
+        DummySqlType::UInt64 => get_and_wrap(context, ValueAccess::as_u64, CValue::UInt64),
+
+        DummySqlType::Int8 => get_and_wrap(context, ValueAccess::as_i8, CValue::Int8),
+
+        DummySqlType::Int16 => get_and_wrap(context, ValueAccess::as_i16, CValue::Int16),
+
+        DummySqlType::Int32 => get_and_wrap(context, ValueAccess::as_i32, CValue::Int32),
+
+        DummySqlType::Int64 => get_and_wrap(context, ValueAccess::as_i64, CValue::Int64),
+
+        DummySqlType::String => get_and_wrap(context, ValueAccess::as_str, |s| {
+            CValue::String(Arc::new(s.as_bytes().to_vec()))
+        }),
+
+        DummySqlType::Ipv4 => convert_string_or_array(
+            context,
+            |ip: Ipv4Addr| ip.octets(),
+            // This is slightly embarrassing...
+            //
+            // ... But it does work!
+            |[a, b, c, d]| CValue::Ipv4([d, c, b, a]),
+            ErrorKind::MalformedIpAddr,
+        ),
+
+        DummySqlType::Ipv6 => convert_string_or_array(
+            context,
+            |ip: Ipv6Addr| ip.octets(),
+            CValue::Ipv6,
+            ErrorKind::MalformedIpAddr,
+        ),
+
+        DummySqlType::Uuid => convert_string_or_array(
+            context,
+            Uuid::into_bytes,
+            CValue::Uuid,
+            ErrorKind::MalformedUuid,
+        ),
+
+        DummySqlType::DateTime => get_and_wrap(context, ValueAccess::as_u32, |timestamp| {
+            CValue::DateTime(timestamp, UTC)
+        }),
+
+        DummySqlType::DateTime64Secs => get_and_wrap(context, ValueAccess::as_i64, |timestamp| {
+            CValue::DateTime64(timestamp, (0, UTC))
+        }),
+
+        DummySqlType::DateTime64Millis => get_and_wrap(context, ValueAccess::as_i64, |timestamp| {
+            CValue::DateTime64(timestamp, (3, UTC))
+        }),
+
+        DummySqlType::DateTime64Micros => get_and_wrap(context, ValueAccess::as_i64, |timestamp| {
+            CValue::DateTime64(timestamp, (6, UTC))
+        }),
+
+        DummySqlType::DateTime64Nanos => get_and_wrap(context, ValueAccess::as_i64, |timestamp| {
+            CValue::DateTime64(timestamp, (9, UTC))
+        }),
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+struct ConversionContext<'a, 'b, 'c, 'd> {
+    column_name: &'a str,
+    value: &'b TValue<'c>,
+    expected_type: &'d DummySqlType,
+}
+
+fn wrap_getter_error<'a, 'b, F, T>(context: ConversionContext<'_, 'a, 'b, '_>, f: F) -> Result<T>
```

Comment:
Could we use more descriptive names for the Type parameters? especially in the functions below they're hard to make out where the `getter` is called `F` and the `wrapper` called `G` it makes the code quite a bit harder to understand

Reply:
Thank you for pointing this out. It should be better as of 32eda5c0.